File: java/base/src/main/java/vsdk/toolkit/io/wrapper/PipeOutputStream.java
package vsdk.toolkit.io.wrapper;

import java.io.IOException;
import java.io.OutputStream;
import java.util.Locale;

public class PipeOutputStream extends OutputStream {
    private Process processHandle;
    private OutputStream pipeOutput;

    private static ProcessBuilder processBuilderForCommand(String command) {
        String osName = System.getProperty("os.name", "").toLowerCase(Locale.ROOT);
        if (osName.contains("win")) {
            return new ProcessBuilder("cmd.exe", "/c", command);
        }
        return new ProcessBuilder("/bin/sh", "-c", command);
    }

    public PipeOutputStream(String command) {
        processHandle = null;
        pipeOutput = null;
        if (command != null && !command.isEmpty()) {
            try {
                ProcessBuilder processBuilder = processBuilderForCommand(command);
                processHandle = processBuilder.start();
                pipeOutput = processHandle.getOutputStream();
            }
            catch (IOException ignored) {
                processHandle = null;
                pipeOutput = null;
            }
        }
    }

    public boolean isOpen() {
        return pipeOutput != null;
    }

    @Override
    public void write(int value) {
        if (pipeOutput == null) {
            return;
        }
        try {
            pipeOutput.write(value & 0xFF);
        }
        catch (IOException ignored) {
        }
    }

    @Override
    public void write(byte[] buffer, int offset, int length) {
        if (pipeOutput == null || buffer == null || offset < 0 || length < 0 || offset + length > buffer.length) {
            return;
        }
        if (length == 0) {
            return;
        }
        try {
            pipeOutput.write(buffer, offset, length);
        }
        catch (IOException ignored) {
        }
    }

    @Override
    public void flush() {
        if (pipeOutput == null) {
            return;
        }
        try {
            pipeOutput.flush();
        }
        catch (IOException ignored) {
        }
    }

    @Override
    public void close() {
        if (pipeOutput == null) {
            return;
        }

        try {
            pipeOutput.flush();
        }
        catch (IOException ignored) {
        }

        try {
            pipeOutput.close();
        }
        catch (IOException ignored) {
        }
        pipeOutput = null;

        if (processHandle != null) {
            processHandle.destroy();
            try {
                processHandle.waitFor();
            }
            catch (InterruptedException e) {
                Thread.currentThread().interrupt();
            }
            processHandle = null;
        }
    }

    public void dispose() {
        close();
    }
}

File: cpp/base/src/main/vsdk/toolkit/io/wrapper/PipeOutputStream.cpp
#include <cstdio>

#include "java/lang/ProcessBuilder.h"
#include "vsdk/toolkit/io/wrapper/PipeOutputStream.h"

FILE *
PipeOutputStream::toFileHandle(void *handle) {
    return static_cast<FILE *>(handle);
}

PipeOutputStream::PipeOutputStream(const char *command):
    pipeHandle(nullptr)
{
    if ( command != nullptr && command[0] != '\0' ) {
        java::ProcessBuilder processBuilder(command);
        pipeHandle = processBuilder.startWrite();
    }
}

PipeOutputStream::~PipeOutputStream() {
    close();
}

bool
PipeOutputStream::isOpen() const {
    return pipeHandle != nullptr;
}

void
PipeOutputStream::write(int value) {
    FILE *handle = toFileHandle(pipeHandle);
    if ( handle == nullptr ) {
        return;
    }
    fputc(static_cast<unsigned char>(value & 0xFF), handle);
}

void
PipeOutputStream::write(const unsigned char *buffer, int offset, int length) {
    FILE *handle = toFileHandle(pipeHandle);
    if ( handle == nullptr || buffer == nullptr || offset < 0 || length < 0 ) {
        return;
    }
    if ( length == 0 ) {
        return;
    }
    fwrite(buffer + offset, 1, static_cast<size_t>(length), handle);
}

void
PipeOutputStream::flush() {
    FILE *handle = toFileHandle(pipeHandle);
    if ( handle == nullptr ) {
        return;
    }
    fflush(handle);
}

void
PipeOutputStream::close() {
    FILE *handle = toFileHandle(pipeHandle);
    if ( handle == nullptr ) {
        return;
    }
    java::ProcessBuilder::close(handle);
    pipeHandle = nullptr;
}

void
PipeOutputStream::dispose() {
    close();
}
